docs/en/02_Developer_Guides/15_Customising_the_Admin_Interface/04_Preview.md

```diff
@@ -7,38 +7,390 @@ summary: How content previews work in the CMS
+page template calls `$MetaTags` in the `<head>` element). When a page loads,
+these meta tags are checked and the appropriate form is loaded.
+
+When rendering a full page in the preview panel to preview a `DataObject` on that
+page, the meta tags for that page are present. When a content author toggles between
+the draft and published preview states, those meta tags are checked and the page's
+edit form would be loaded instead of the `DataObject`'s form. To resolve this, you
+can include an extra GET parameter in the `PreviewLink`. Then in `MetaTags`
+when the extra parameter is detected, omit the relevant meta tags.
+
+Note that this is not necessary for unversioned `DataObjects` as they only have
+one preview state.
+
+```php
+use SilverStripe\Control\Controller;
+use SilverStripe\View\Parsers\HTML4Value;
+
+class ProductPage extends Page
+{
+    //...
+
+    private static $has_many = [
+        'Products' => Product::class,
+    ];
+
+    public function MetaTags($includeTitle = true)
+    {
+        $tags = parent::MetaTags($includeTitle);
+        if (!Controller::has_curr()) {
+            return;
+        }
+        // If the 'DataObjectPreview' GET parameter is present, remove 'x-page-id' and 'x-cms-edit-link' meta tags.
+        // This ensures that toggling between draft/published states doesn't revert the CMS to the page's edit form.
+        $controller = Controller::curr();
+        $request = $controller->getRequest();
+        if ($request->getVar('DataObjectPreview') !== null) {
+            $html = HTML4Value::create($tags);
+            $xpath = "//meta[@name='x-page-id' or @name='x-cms-edit-link']";
+            $removeTags = $html->query($xpath);
+            $body = $html->getBody();
+            foreach ($removeTags as $tag) {
+                $body->removeChild($tag);
+            }
+            $tags = $html->getContent();
+        }
+        return $tags;
+    }
+}
+```
+
+#### The DataObject Implementation
+Make sure the Versioned `Product` class implements `CMSPreviewable` and enables
+the draft and published preview states.
+
+```php
+use SilverStripe\ORM\CMSPreviewable;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\Versioned\Versioned;
+
+class Product extends DataObject implements CMSPreviewable
+{
+    private static $show_stage_link = true;
+    private static $show_live_link = true;
+
+    private static $extensions = [
+        Versioned::class,
+    ];
+
+    private static $has_one = [
+        'Page' => Page::class,
+    ];
+
+    // ...
+
+    public function PreviewLink($action = null)
+    {
+        return null;
+    }
+
+    public function CMSEditLink()
+    {
+        return null;
+    }
+
+    public function getMimeType()
+    {
+        return 'text/html';
+    }
+
+}
+```
+
+Implement a method which gives you a unique repeatable anchor for each
+distinct `Product` object.
+
+```php
+/**
+ * Used to generated the id for the holder element in the template.
```

Comment:
```suggestion
 * Used to generated the id for the product element in the template.
```
Oops, this was a copy/paste error. This concept was taken from the elemental module so there may be a couple other holdovers from that which I missed.